Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp

```diff
@@ -0,0 +1,386 @@
+        decoder->close();
+    });
+}
+
+Expected<std::pair<FourCharCode, std::optional<AudioStreamDescription::PCMFormat>>, String> AudioDecoderCocoa::isCodecSupported(const StringView& codecName)
+{
+    auto codec = [](auto& codecName) -> FourCharCode {
+        if (codecName.startsWith("mp4a.40.5"_s))
+            return kAudioFormatMPEG4AAC;
+        if (codecName.startsWith("mp4a.40.23"_s))
+            return kAudioFormatMPEG4AAC_LD;
+        if (codecName.startsWith("mp4a.40.29"_s))
+            return kAudioFormatMPEG4AAC_HE_V2;
+        if (codecName.startsWith("mp4a.40.39"_s))
+            return kAudioFormatMPEG4AAC_ELD;
+        if (codecName.startsWith("mp4a"_s))
+            return kAudioFormatMPEG4AAC;
+        if (codecName == "mp3"_s)
+            return kAudioFormatMPEGLayer3;
+        if (codecName == "opus"_s)
+            return kAudioFormatOpus;
+        if (codecName == "alaw"_s)
+            return kAudioFormatALaw;
+        if (codecName == "ulaw"_s)
+            return kAudioFormatULaw;
+        if (codecName == "flac"_s)
+            return kAudioFormatFLAC;
+        if (codecName == "vorbis"_s)
+            return 'vorb';
+        if (codecName.startsWith("pcm-"_s))
+            return kAudioFormatLinearPCM;
+        return 0;
+    }(codecName);
+
+    if (!codec)
+        return makeUnexpected("not supported codec"_s);
+
+    std::optional<AudioStreamDescription::PCMFormat> format;
+    if (codec != kAudioFormatLinearPCM)
+        return std::make_pair(codec, format);
+
+    auto components = codecName.toString().split('-');
+    if (components.size() != 2)
+        return makeUnexpected(makeString("Invalid LPCM codec string:"_s, codecName));
+    auto pcmFormat = components[1].convertToASCIILowercase();
+    codec = kAudioFormatLinearPCM;
+    if (pcmFormat == "u8"_s)
+        format = AudioStreamDescription::Uint8;
+    else if (pcmFormat == "s16"_s)
+        format = AudioStreamDescription::Int16;
+    else if (pcmFormat == "s24"_s)
+        format = AudioStreamDescription::Int24;
+    else if (pcmFormat == "s32"_s)
+        format = AudioStreamDescription::Int32;
+    else if (pcmFormat == "f32"_s)
+        format = AudioStreamDescription::Float32;
+    else
+        return makeUnexpected(makeString("Invalid LPCM codec format:"_s, pcmFormat));
+
+    return std::make_pair(codec, format);
+}
+
+Ref<AudioDecoder::DecodePromise> AudioDecoderCocoa::decode(EncodedData&& data)
+{
+    return invokeAsync(queueSingleton(), [data = SharedBuffer::create(data.data), isKeyFrame = data.isKeyFrame, timestamp = data.timestamp, duration = data.duration, decoder = m_internalDecoder]() mutable {
+        return decoder->decode(WTFMove(data), isKeyFrame, timestamp, duration);
+    });
+}
+
+Ref<GenericPromise> AudioDecoderCocoa::flush()
+{
+    return invokeAsync(queueSingleton(), [decoder = m_internalDecoder] {
+        return decoder->flush();
+    });
+}
+
+void AudioDecoderCocoa::reset()
+{
+    // FIXME: that can't be correct
```

Comment:
The goal of reset to set closed synchronously was to limit the number of decode callback calls.
It should not be observable to dispatch to the work queue.

The FIXME could probably be removed.